### PR TITLE
Adding data url export and save/load with LZ-String compression features

### PR DESCRIPTION
**Pull request by filipemeneses**
Hey, I added these features, it's interesting if you need to store the configuration

### CODE REVIEW

**Comment by stagas** [comment]
Woah, this is super cool!! Thank you so much for contributing, I appreciate the effort! I am very tempted to hit the Merge button but I'm thinking that it's a too specific use case we are solving.. maybe we can achieve the same thing a different way:

The `.export()` function basically converts to canvas, so maybe we can have a `.toCanvas()` method then implement `.export()` on top of that, just to reduce verbosity:

```js
PixelArt.prototype.export = function(filetype, ratio) {
  return this.toCanvas().toDataURL(filetype, ratio)
}
```

Similarly, and instead of binding the user to a single compression method, we can have a `.toString()` and accept that output as the constructor input:

```js
var lz = LZString.compressToUint8Array(nyan.toString())
...
var nyan = new PixelArt(LZString.decompressFromUint8Array(lz))
```

This way we achieve the functionality we need without binding the user to a specific use case (and removing the `LZString` dependency from the build when it's not used). What are your thoughts?

**Comment by filipemeneses** [comment]
Oh, nice idea! I tested a few text compressors to pick LZString, but you're right, forcing a single compression method we're creating a limitation. I even thought of creating a dictionary on top of LZString, but then we would end up with a new type of compression and also limiting it.

I've made that string format to work with LZString compression, but, in this case, we might choose `.toJSON()` or `.toObject()` with will output all the configurations and let the developer decides what fits best.

```js
var nyanJson = nyan.toJSON()
// <- {art, palette } 
```

And then the developer decides how to build the string, giving him more liberty. What do you think?

I'll make the `.toCanvas()` adjusts.

**Comment by stagas** [comment]
I agree! The `.toJSON()` method can output all pieces serialized (even pos/scale) in an object as you describe. Also `JSON.stringify(nyan)` would work out of the box this way, so serializing a hypothetical `world` object that contains many `PixelArt` instances would serialize all of them properly. The only issue I see is that the resulting string would also contain the object keys so it wouldn't compress so well, which defeats your original use case I think. Any idea how to solve this and still keep the API simple?

Obviously, we would also need to make `new PixelArt(JSON.parse(JSON.stringify(nyan)))` work as well. And, by including pos/scale in the serialization, means we can effortlessly share complete compositions over the network or save to disk and reproduce exactly on retrieve.

What do you think?

**Comment by filipemeneses** [comment]
I see, you're right. If we want to serialize anything, we can easily do:
```js
nyan = JSON.stringify(nyan)
// <-- String
nyan = Object.assign(Object.create(new PixelArt), JSON.parse(nyan))
// <-- A new clone
```
So you don't need to implement anything to parse: `new PixelArt(JSON.parse(JSON.stringify(nyan))) `.

The `.toJSON()` idea is to give the developer all the relevant data from that art within an organized object. So, we're going to need a `.fromJSON()` if we want to keep the instance.

The idea of the string was to come up with only the essentials (without any object keys), like the `art` and `palette`. I realize I made a mistake by using `JSON.stringify(palette)`, so, maybe coming up with something smarter for serializing the palette will reduce the string size considerably. 

So:

- implementing `.toJSON` and `.fromJSON` is an easy way to save and restore the PixelArt, but will consume more space and less processing.
- implementing `.toString` and `.fromString` is an easy way to save and restore the PixelArt, but will consume more processing and less space.
 
And that `.toString` should output a very compressed string, not just `JSON.stringify`.

**Comment by stagas** [comment]
Umm ok, yeah since we can serialize using `JSON.stringify(nyan)` and we are not solving any existing use case with a `.toJSON()`, I propose we leave it out for the time being and just implement `.toString()` and `.fromString()`, basically the same as `.save()` and `.load()` are right now but without the LZ compression, plus the `.toCanvas()` with the `.export()` helper. So at least you solve your use case and are unblocked and the API is cleaner / more versatile this way. And when we come up with a better compression for the palette we can just add it then, the API remains the same anyway so the change should be transparent.

**Comment by filipemeneses** [comment]
Done. Here's what we got:

- `.toString()` Made a simple compressor on `toString()` that remove the repetitions, like: `aaaaaaaaaa` becomes `a10`. This will cause a limitation where there can't be numbers as "pixels". Each line separated with `{}` and the palette by a `[]`. Funny thing: The last `nyan` compression reduced to 1207 characters, this one reduced to 504.
- `.fromString()` Made a example to make sure it works.
- `.toCanvas()` return canvas element
- `.export()` return DataUrl

File: Readme.md [hunk 1]
```diff
@@ -59,6 +59,45 @@ Result:
...
+## Save
+
+To save the compressed art string, use:
+
+```js
+nyan.save()
+// <- Compressed string with art and palette
+```
+
+> It uses [LZ-String](https://github.com/pieroxy/lz-string/). In the example, it compressed 25% more compared to the image/webp string compression
```
Comment (stagas) [inline Readme.md]: No longer valid, correct?

**Comment by stagas** [comment]
Looks good. I dig the compression algorithm you came up with! A couple of things only: 

1) We could use `\n` for rows delimiter and `\t` for palette or you want to keep visible characters for some reason?

2) The file permissions changed from 644 to 755 which is not right as they become executables, probably was by accident.

File: Readme.md (hunk 1 above)
Comment (filipemeneses) [reply Readme.md]: Oh, my bad. I'll update it.

**Comment by filipemeneses** [comment]
1. I firstly pick `\n` and `\t` as delimiters, but as being hidden characters, the developer might have some problem with it while handling the strings. (not programmatically, of course, just a bad copy and paste)

2. Hm, I'll give a look into that. Switched from Ubuntu to MacOS. No idea what happened.

**Comment by filipemeneses** [comment]
There we go, fixed the file permissions issue and removed completely any track of LZ-String. haha

**Comment by filipemeneses** [comment]
Can you npm publish this pull please?

**Comment by stagas** [comment]
Published as v1.3.0 on npm.

**Pull request merged**